subgraphs/osmosis/src/modules/Swap.ts

```diff
@@ -0,0 +1,404 @@
+
+  for (let i = 0; i < attributes.length; i++) {
+    const key = attributes[i].key;
+    const value = attributes[i].value;
+    if (key == "pool_id") {
+      poolId = value;
+    } else if (key == "tokens_in" || key == "tokens_out") {
+      const j = tokenDataParser(value) as i32;
+      let tokenAmount = BigInt.fromString(value.substring(0, j));
+      const tokenDenom = value.substring(j, value.length);
+      if (key == "tokens_in") {
+        tokenInAmount = tokenAmount;
+        tokenInDenom = tokenDenom;
+      } else {
+        tokenOutAmount = tokenAmount;
+        tokenOutDenom = tokenDenom;
+      }
+    }
+  }
+
+  const liquidityPoolId = constants.Protocol.NAME.concat("-").concat(poolId);
+  const liquidityPool = LiquidityPoolStore.load(liquidityPoolId);
+  if (liquidityPool == null) {
+    return;
+  }
+
+  log.warning(
+    "swap() at height {} index {} tokenInDenom {} tokenInAmount {}  tokenOutDenom {} tokenOutAmount {} ",
+    [
+      tx.height.toString(),
+      tx.index.toString(),
+      tokenInDenom.toString(),
+      tokenInAmount.toString(),
+      tokenOutDenom.toString(),
+      tokenOutAmount.toString(),
+    ]
+  );
+  swap(
+    poolId,
+    tokenInAmount,
+    tokenInDenom,
+    tokenOutAmount,
+    tokenOutDenom,
+    data,
+    tx,
+    block
+  );
+}
+
+function tokenDataParser(tokenData: string): number {
+  let i = 0;
+  while (i < tokenData.length) {
+    const tokeChar = tokenData.charAt(i);
+    if (
+      tokeChar == "0" ||
+      tokeChar == "1" ||
+      tokeChar == "2" ||
+      tokeChar == "3" ||
+      tokeChar == "4" ||
+      tokeChar == "5" ||
+      tokeChar == "6" ||
+      tokeChar == "7" ||
+      tokeChar == "8" ||
+      tokeChar == "9"
+    ) {
+      i++;
+    } else {
+      break;
+    }
+  }
+
+  return i;
+}
+
+function swap(
+  poolId: string,
+  tokenInAmount: BigInt,
+  tokenInDenom: string,
+  tokenOutAmount: BigInt,
+  tokenOutDenom: string,
+  data: cosmos.Event,
+  tx: cosmos.TxResult,
+  block: cosmos.HeaderOnlyBlock
+): void {
+  const liquidityPoolId = constants.Protocol.NAME.concat("-").concat(poolId);
+  const liquidityPool = LiquidityPoolStore.load(liquidityPoolId);
+  if (liquidityPool == null) {
+    return;
+  }
+
+  const inputTokenBalances = liquidityPool.inputTokenBalances;
+  const inputTokenAmounts = new Array<BigInt>(inputTokenBalances.length).fill(
+    constants.BIGINT_ZERO
+  );
+  const inputTokens = liquidityPool.inputTokens;
+  const tokenInIndex = inputTokens.indexOf(tokenInDenom);
+  if (tokenInIndex >= 0) {
+    inputTokenBalances[tokenInIndex] = inputTokenBalances[tokenInIndex].plus(
+      tokenInAmount
+    );
+    inputTokenAmounts[tokenInIndex] = tokenInAmount;
+  }
+
+  const tokenOutIndex = inputTokens.indexOf(tokenOutDenom);
+  if (tokenOutIndex >= 0) {
+    inputTokenBalances[tokenOutIndex] = inputTokenBalances[tokenOutIndex].minus(
+      tokenOutAmount
+    );
+    if (inputTokenBalances[tokenOutIndex] <= constants.BIGINT_ZERO) {
```

Comment:
Why do you return here can you put a note on this?

Reply:
Add a note and log here at https://github.com/messari/subgraphs/commit/6903f52ce28319ef61b2eb05dafdabf73a47568c